### ISSUE TITLE
unhandled OSError in get_mount_facts for linux

### ISSUE THREAD

**Issue by thatmattbone** [ansible/ansible#3183]
I use ansible as a non-superuser and it seems like a change gathering facts in ansible 1.2 results in an OSError. This is because I don't have permission to call `os.statvfs` on something listed in `/etc/mtab`. Everything works as expected in ansible 1.1.

The offending change is:
https://github.com/ansible/ansible/commit/73a12b04a4e69248576807d148feeeb167c78d65

It's easy enough to revert this change or modify the code to ignore the error, but I don't know how that'd impact things more broadly.

**Comment by mpdehaan** [comment]
What OS are you testing on?

**Comment by thatmattbone** [comment]
CentOS 6.4. 

Commenting out the call to `os.statvfs` (and removing those items from the mounts dict) fixes the problem for me.

Here's the whole traceback:

```
fatal: [192.168.56.121] => failed to parse: Traceback (most recent call last):
  File "/tmp/ansible-1371126080.08-161579759134734/setup", line 2906, in <module>
    main()
  File "/tmp/ansible-1371126080.08-161579759134734/setup", line 2008, in main
    data = run_setup(module)
  File "/tmp/ansible-1371126080.08-161579759134734/setup", line 1947, in run_setup
    facts = ansible_facts()
  File "/tmp/ansible-1371126080.08-161579759134734/setup", line 1937, in ansible_facts
    facts.update(Hardware().populate())
  File "/tmp/ansible-1371126080.08-161579759134734/setup", line 435, in populate
    self.get_mount_facts()
  File "/tmp/ansible-1371126080.08-161579759134734/setup", line 550, in get_mount_facts
    statvfs_result = os.statvfs(fields[1])
OSError: [Errno 13] Permission denied: '/var/named/chroot/etc/named'
```

`/var/named/chroot/etc/named` is listed in `/etc/mtab` but unfortunately I have to run ansible as a user that doesn't have permissions to stat that file.

**Comment by mikesac** [comment]
getting the same with ansible version 1.2.1 , master machine is ubuntu, destination CentOs, using unprivileged user as well.
Any workaround on this ?

Update: disabling fact gathering (gather_facts: False), skips the problem, but then I'm missing all fact-related variables like ${ansible_user_id} :(

**Comment by ingmar** [comment]
Same problem here (the mount point in question is a pg_xlog and therefore has intentionally restrictive permissions). We ssh into hosts as user and sudo where and when needed.

I see two ways of resolving this (not mutually exclusive):
- The fact getter could have an option to run sudo.
- **get_mount_facts** should catch the Permission denied OSError and mark the statvfs values as unavailable instead of the whole playbook falling on its nose.

**Comment by mpdehaan** [comment]
Can you please try this in Ansible 1.3?

**Comment by mpdehaan** [comment]
checkout latest development branch

source ./hacking/env-setup

that will get you running from checkout without an installation

**Comment by ingmar** [comment]
Works in c222c8133bad91f32067fd77b059f57bcd25ba29. I just hacked a workaround for 1.2 meanwhile:

```
--- setup.orig  2013-07-16 15:58:48.000000000 +0100
+++ setup   2013-07-16 15:58:52.000000000 +0100
@@ -547,16 +547,27 @@
         for line in mtab.split('\n'):
             if line.startswith('/'):
                 fields = line.rstrip('\n').split()
-                statvfs_result = os.statvfs(fields[1])
+                try:
+                    statvfs_result = os.statvfs(fields[1])
+                except OSError as e:
+                    if e.errno == errno.EACCES:
+                        statvfs_result = None
+                    else:
+                        raise
+                else:
+                    continue
                 self.facts['mounts'].append(
                     {'mount': fields[1],
                      'device':fields[0],
                      'fstype': fields[2],
                      'options': fields[3],
-                     # statvfs data
-                     'size_total': statvfs_result.f_bsize * statvfs_result.f_blocks,
-                     'size_available': statvfs_result.f_bsize * (statvfs_result.f_bavail),
                      })
+                if statvfs_result is not None:
+                    self.facts['mounts'].append(
+                        {
+                         'size_total': statvfs_result.f_bsize * statvfs_result.f_blocks,
+                         'size_available': statvfs_result.f_bsize * (statvfs_result.f_bavail),
+                         })

     def get_device_facts(self):
         self.facts['devices'] = {}
```

**Comment by mikesac** [comment]
I can confirm latest devel is working properly, thanks a lot! 
Will this be released as 1.2.2 ?

**Comment by mpdehaan** [comment]
Thanks!

1.2.2 is already released and I am not planning another dot release at this time.
